build-utils/mount-partitions.sh

```diff
@@ -0,0 +1,100 @@
+# from the end so we can compare with the base device.
+current_boot_device=$(grep /etc/resolv.conf /proc/mounts | awk '{print substr($1,1,length($1)-2)}')
+if [ "${current_boot_device}" = "" ]; then
+    echo "WARNING: Could not determine boot device. Please launch Supervisor as a privileged container with host networking."
+fi
+
+# Mounts a device to a path if it's not already mounted.
+# Usage: do_mount DEVICE MOUNT_PATH 
+do_mount() {
+    device=$1
+    path=$2
+
+    # Create the directory if it doesn't exist
+    mkdir -p "${path}"
+
+    # Mount the device if it doesn't exist
+    if [ "$(mountpoint -n "${path}" | awk '{ print $1 }')" != "${device}" ]; then
+        mount "${device}" "${path}"
+    fi
+}
+
+# Find the devices for each balenaOS partition.
+# Usage: setup_then_mount PARTITION MOUNT_PATH
+# PARTITION should be one of boot, state, or data. 
+setup_then_mount() {
+    # If in test environment, pretend we've succeeded at mounting everything to their
+    # new mountpoints. We don't want to actually mount in a containerized test environment 
+    # where the Supervisor is probably not running on a host that has the needed partitions.
+    if [ "${TEST}" = 1 ]; then
+        return 0
+    fi
+
+    partition=$1
+    path=$2
+
+    # Get one or more devices matching label, accounting for legacy partition labels.
+    device=$(blkid | grep -E "(resin|balena)-${partition}" | awk -F':' '{print $1}')
+    
+    count=$(echo "${device}" | grep -c "/dev")
+    if [ "${count}" = 0 ]; then
+        # If no devices with the partition label are found, use legacy mounts.
+        echo "WARNING: No partitions labeled '(resin|balena)-${partition}' found. $use_legacy_message"
+        return 1
+    fi
+
+    if [ "${count}" = 1 ]; then
+        # QUESTION: Should we check whether do_mount succeeded in mounting and inform of failure?
+        do_mount "${device}" "${path}"
+        return 0
+    fi
+
+    # If multiple devices with the partition label are found, mount to the device
+    # that's part of the current boot device. as this indicates a duplicate 
+    # label somewhere created by a user or an inconsistency in the system.
+    # We've been able to identify the current boot device, so use that
+    # to find the device with the correct label amongst 2+ devices.
+    found=false
+    for d in ${device}; do
+        if [ "$(echo "$d" | grep -E "^$current_boot_device")" != "" ]; then
+            echo "INFO: Found device $d on current boot device $current_boot_device, using as mount for '(resin|balena)-${partition}'."
+            do_mount "${d}" "${path}"
+            found=true
+        fi
+    done
+
+    if [ ! $found ]; then
+        echo "WARNING: Multiple partitions labeled '(resin|balena)-${partition}' found but none from current boot device. $use_legacy_message"
+        return 1
+    fi
+
+    return 0
+}
+
+# Set overlayfs root mountpoint
+export ROOT_MOUNTPOINT="/mnt/root"
```

Comment:
I'm still doubtful on keeping the `/mnt/root` fallback. I agree it might be useful while we ensure to have covered all border cases, but we probably should stop relying on it. If a mounting step fails, the system may have bigger problems that might even become worse by starting a supervisor.

Reply:
The eventual goal would be to remove this, but we can't merge this change today without the `/mnt/root` dependency due to needing it for update locks and checking the VPN status.

Reply:
Never mind, I think I know what you mean -- we don't want to fall back to /mnt/root/mnt/boot at all if the partitions aren't found, right? Just exit with error?

Reply:
yeah, that’s what I was thinking of

Reply:
Removed the fallback to ROOT_MOUNTPOINT as base path for boot, data, and state partitions.